### PR TITLE
Add FIPS Selftest module integrity check

### PR DESCRIPTION
**Pull request by slontis**
##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by slontis** [comment]
@levitte - I am wondering what to do with test_evp_extra inside the fetch tests that are currently doing..

/*
 * Test 0: Test with the default OPENSSL_CTX
 * Test 1: Test with an explicit OPENSSL_CTX
 * Test 2: Explicit OPENSSL_CTX with explicit load of default provider
 * Test 3: Explicit OPENSSL_CTX with explicit load of default and fips provider
 * Test 4: Explicit OPENSSL_CTX with explicit load of fips provider
 */
static int test_EVP_MD_fetch(int tst)..

The issue I have is 
1) the auto config file loading gets done once and all of these tests use different config files" - Do we need the config to be part of the provider load to do this.
2) The current behaviour is to use the default provider if nothing is loaded, but this means it also does this when an explicit load of the fips provider fails. (So it passes - when it should probably fail).

File: test/evp_extra_test.c [hunk 1]
```diff
@@ -1119,6 +1120,13 @@ static int test_EVP_MD_fetch(int tst)
         if (!TEST_ptr(ctx))
             goto err;
 
+        if (tst == 2)
+            setenv("OPENSSL_CONF", "test/default.cnf", 1);
+        if (tst == 3)
+            setenv("OPENSSL_CONF", "test/default-and-fips.cnf", 1);
+        if (tst == 4)
+            setenv("OPENSSL_CONF", "fips.cnf", 1);
```
Comment (levitte) [inline test/evp_extra_test.c]: It's possible this might not work the way you think on certain platforms. I will have to check...

Comment (slontis) [reply test/evp_extra_test.c]: Even if it does is this really what we want to do in the test?

Comment (levitte) [reply test/evp_extra_test.c]: Oh, you did start a discussion on that...  sorry, that blew right past me for some reason.

**Comment by levitte** [comment]
It could be that we should split `evp_extra_test.c` in smaller more dedicated programs, *or* parametrize it to perform different test suites depending on command line arguments (like shlibloadtest does)

**Comment by slontis** [comment]
Ok just checking that something didnt need to be changed in the API.. will look into doing that.
The only other issue then will be related to the fallback plan it uses when the autoload of the fips module fails.. (Somewhere in the config it has some flags that just say ignore any modules that dont load properly - so the default gets loaded).

**Comment by slontis** [comment]
@levitte - the failure is due to me using .include ../fips_install.cnf inside the test fips.cnf? Can you think of a way around this without having to generate the files on the fly? the travis config for install doesnt like this path too much..

**Comment by levitte** [comment]
If I'm reading the `.include` implementation right, all inclusions are relative the current working directory of the process being run.

Either way, *temporary test files should be created in the build directory*, not the source directory.

File: test/recipes/30-test_evp.t [hunk 2]
```diff
@@ -27,7 +33,16 @@ my @defltfiles = qw( evpencod.txt evpkdf.txt evppkey_kdf.txt evpmac.txt
     evppbe.txt evppkey.txt evppkey_ecc.txt evpcase.txt evpaessiv.txt
     evpccmcavs.txt );
 
-plan tests => (scalar(@configs) * scalar(@files)) + scalar(@defltfiles);
+plan tests => (scalar(@configs) * scalar(@files)) + scalar(@defltfiles) + 1;
+
+my $infile = bldtop_file('providers', platform->dso('fips'));
+$ENV{OPENSSL_MODULES} = bldtop_dir("providers");
+
+ok(run(app(['openssl', 'fipsinstall', '-out', srctop_file('test', 'fipsinstall.conf'),
```
Comment (levitte) [inline test/recipes/30-test_evp.t]: Make that `bldtop_file`

File: test/recipes/30-test_evp_fetch_prov.t [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+   ok(run(test(["evp_fetch_prov_test", "-type", "$alg", "-property", "default=yes", "default"])),
+       "running evp_fetch_prov_test with $alg fetch 'default=yes' using default provider loaded");
+   ok(run(test(["evp_fetch_prov_test", "-type", "$alg", "-property", "fips=no", "default"])),
+       "running evp_fetch_prov_test with $alg fetch 'fips=no' using default provider loaded"); 
+   ok(run(test(["evp_fetch_prov_test", "-type", "$alg", "-property", "default=no", "-fetchfail", "default"])),
+       "running evp_fetch_prov_test with $alg fetch 'default=no' using default provider loaded should fail");
+   ok(run(test(["evp_fetch_prov_test", "-type", "$alg", "-property", "fips=yes", "-fetchfail", "default"])),
+       "running evp_fetch_prov_test with $alg fetch 'fips=yes' using default provider loaded should fail");
+   
+   $ENV{OPENSSL_CONF} = srctop_file("test", "fips.cnf");
```
Comment (levitte) [inline test/recipes/30-test_evp_fetch_prov.t]: I believe you've made this test depend on `30-test_evp.t`, which is a big no-no.  What if someone only does `make test TESTS=test_evp_fetch_prov`?

The test recipes must be independent from each other

File: test/default-and-fips.cnf [hunk 4]
```diff
@@ -0,0 +1,13 @@
+openssl_conf = openssl_init
+
+.include ../fipsinstall.conf
```
Comment (levitte) [inline test/default-and-fips.cnf]: I suspect that these need to be generated.  What you can do is this:

-   Add the following line in `test/build.info`:

    ```
    DEPEND[]=default-and-fips.cnf fips.cnf
    SOURCE[default-and-fips.cnf]=default-and-fips.cnf.in
    SOURCE[fips.cnf]=fips.cnf.in
    ```
-   rename those two .cnf files to .cnf.in, and change the .include like this:

    ``` perl
    .include {- use Cwd 'abs_path'; use File::Spec::Functions;
                abs_path(catfile($config{builddir}, 'test', 'fipsinstall.conf')); -}
    ```

**Comment by levitte** [comment]
The alternative to what I'm suggesting is that the `.include` support must be re-worked to become aware of the directory of the including file and include relative to that.

**Comment by t8m** [comment]
> The alternative to what I'm suggesting is that the `.include` support must be re-worked to become aware of the directory of the including file and include relative to that.

That would be quite non-trivial as a configuration file can be loaded from arbitrary BIO. Which may not be even a file. Maybe add some API call to set the root directory for the relative .include paths?

File: test/recipes/30-test_evp_fetch_prov.t (hunk 3 above)
Comment (slontis) [reply test/recipes/30-test_evp_fetch_prov.t]: Was aware of the issue - it regenerates the data for this test also

File: test/default-and-fips.cnf (hunk 4 above)
Comment (slontis) [reply test/default-and-fips.cnf]: I have gone with using an environment variable instead -  I think this makes it a bit more flexible for installation.

Comment (slontis) [reply test/default-and-fips.cnf]: We could add a API also if required at a later date..

File: doc/man5/config.pod [hunk 5]
```diff
@@ -40,7 +40,8 @@ It is strongly recommended to use absolute paths with the B<.include>
 directive. Relative paths are evaluated based on the application current
 working directory so unless the configuration file containing the
 B<.include> directive is application specific the inclusion will not
-work as expected.
+work as expected. The environment variable B<OPENSSL_CONF_INCLUDE> can also be
+used to specify the path for all .include files.
```
Comment (t8m) [inline doc/man5/config.pod]: Maybe more precise: "... can also be used to specify the path that is prepended to all .include paths."

File: doc/man5/config.pod [hunk 6]
```diff
@@ -487,6 +488,10 @@ Ignored in set-user-ID and set-group-ID programs.
 The path to the directory with OpenSSL modules, such as providers.
 Ignored in set-user-ID and set-group-ID programs.
 
+=item B<OPENSSL_CONF_INCLUDE>
+
+The optional path to prepend to any file included using .include.
```
Comment (t8m) [inline doc/man5/config.pod]: s/file/path/ ?

File: crypto/conf/conf_def.c [hunk 7]
```diff
@@ -360,17 +362,34 @@ static int def_load_bio(CONF *conf, BIO *in, long *line)
                 trim_ws(conf, p);
                 if (!str_copy(conf, psection, &include, p))
                     goto err;
+
+                if (include_dir != NULL) {
+                    size_t newlen = strlen(include_dir) + strlen(include) + 2;
+
+                    include_path = OPENSSL_malloc(newlen);
+                    OPENSSL_strlcpy(include_path, include_dir, newlen);
+                    OPENSSL_strlcat(include_path, "/", newlen);
```
Comment (t8m) [inline crypto/conf/conf_def.c]: Should this "/" append be done only if there is no '/' at the end of include_dir or beginning of include?

File: doc/man5/config.pod (hunk 6 above)
Comment (slontis) [reply doc/man5/config.pod]: changed to The optional path to prepend to all .include paths.

File: crypto/conf/conf_def.c (hunk 7 above)
Comment (slontis) [reply crypto/conf/conf_def.c]: I copied this pattern from elsewhere.. Would it not work if there was an extra / ?

Comment (t8m) [reply crypto/conf/conf_def.c]: It would work fine. It's just ugly.

Comment (slontis) [reply crypto/conf/conf_def.c]: So my initial thought was that the check would also be a bit ugly - (And potentially need to deal with platform specific path seperators).

Comment (levitte) [reply crypto/conf/conf_def.c]: I see a different issue with this code.  Say we have these two conf files:

/tmp/blah.cnf:
```
.include foo/something.cnf
```

/tmp/foo/something.cnf:
```
.include ../hickup.cnf
```

What hickup.cnf do you expect to be included, /tmp/hickup.cnf or /hickup.cnf?

I compare this to what we have learned to expect from the C preprocessor re relative inclusions

Comment (levitte) [reply crypto/conf/conf_def.c]: (not your fault, @slontis, the code was broken in that regard before you touched it, I know that)

Comment (slontis) [reply crypto/conf/conf_def.c]: You perl docs recommend not using relative paths....
This seems to be a different issue to prepending a path to the existing path..

Comment (levitte) [reply crypto/conf/conf_def.c]: *My* perl docs?
Yes, it's a different issue, it's true...  and I had completely forgotten the caveat in config(5).
So err, ok, relegated to another PR, message received...

**Comment by slontis** [comment]
Rebased again..

**Comment by slontis** [comment]
I discovered some 'leftovers' in the new test files.. So cleaned them up also..

**Comment by slontis** [comment]
Thanks for reviewing. Merged to master.

**Pull request closed without merging**